src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectConverterFactory.cs

```diff
@@ -39,11 +40,18 @@ public override bool CanConvert(Type typeToConvert)
             Justification = "The ctor is marked RequiresUnreferencedCode.")]
         public override JsonConverter CreateConverter(Type typeToConvert, JsonSerializerOptions options)
         {
+            ConstructorInfo? constructor;
             JsonConverter converter;
             Type converterType;
 
-            bool useDefaultCtorInUnannotatedStructs = _useDefaultConstructorInUnannotatedStructs && !typeToConvert.IsKeyValuePair();
-            if (!typeToConvert.TryGetDeserializationConstructor(useDefaultCtorInUnannotatedStructs, out ConstructorInfo? constructor))
+            if (typeToConvert.IsKeyValuePair())
+            {
+                // browser-wasm compat -- ensure the linker doesn't trim away constructor parameter names from KVP.
+                Type[] genericArguments = typeToConvert.GetGenericArguments();
+                Type keyValuePairType = typeof(KeyValuePair<,>).MakeGenericType(genericArguments);
```

Comment:
@eerhardt the code in line 51 is redundant, but it's designed to have the linker not trim away the constructor parameter names of KeyValuePair. Do you know of any incantations that might let us avoid it in this case?

Reply:
I believe if you put a `[DynamicDependency(DynamicallyAccessedMemberTypes.PublicConstructors, typeof(KeyValuePair<,>)]` on this method, the trimmer will see that you are going to access the constructor through reflection and not trim away the parameter names.

cc @vitek-karas

Reply:
Good point, we are doing that already to get tests to work in Blazor:

https://github.com/dotnet/runtime/blob/0570583829c7656778c2b0e8f20570627d6ecde1/src/libraries/System.Text.Json/tests/Common/ConstructorTests/ConstructorTests.ParameterMatching.cs#L509-L513

I'd be inclined to just put that annotation in the relevant tests rather than arbitrarily doing it for KVP in product code.

Reply:
> I'd be inclined to just put that annotation in the relevant tests rather than arbitrarily doing it for KVP in product code.

Agreed, especially since this code is marked "RequiresUnreferencedCode". That is the stance we've taken elsewhere in STJ w.r.t. trimming and our "unsafe" APIs. For example - https://github.com/dotnet/runtime/pull/53317